src/c++/perf_analyzer/infer_data_manager.h

```diff
@@ -0,0 +1,224 @@
+  std::unique_ptr<uint8_t, std::function<void(uint8_t*)>> data_;
+};
+
+// Manages infer data to prepare an inference request and the resulting
+// inference output from triton server
+class InferDataManager {
+ public:
+  InferDataManager(
+      const int32_t batch_size, const SharedMemoryType shared_memory_type,
+      const size_t output_shm_size, const std::shared_ptr<ModelParser>& parser,
+      const std::shared_ptr<cb::ClientBackendFactory>& factory,
+      const std::shared_ptr<DataLoader>& data_loader);
+
+  virtual ~InferDataManager();
+
+  /// Helper function to allocate and prepare shared memory.
+  /// \return cb::Error object indicating success or failure.
+  cb::Error InitInferDataManager();
+
+
+  /// Creates inference input object
+  /// \param infer_input Output parameter storing newly created inference input
+  /// \param kind Backend kind
+  /// \param name Name of inference input
+  /// \param dims Shape of inference input
+  /// \param datatype Data type of inference input
+  /// \return cb::Error object indicating success or failure.
+  virtual cb::Error CreateInferInput(
+      cb::InferInput** infer_input, const cb::BackendKind kind,
+      const std::string& name, const std::vector<int64_t>& dims,
+      const std::string& datatype);
+
+  /// Wrapper function to prepare the InferData for sending
+  /// inference request.
+  /// \param ctx The target InferData object.
+  /// \return cb::Error object indicating success or failure.
+  cb::Error PrepareInfer(InferData& ctx);
+
+
+  /// Updates the expected output data to use for inference request. Empty
+  /// vector will be returned if there is no expected output associated to the
+  /// step.
+  /// \param outputs The vector of outputs to get the expected data
+  /// \param stream_index The data stream to use for next data
+  /// \param step_index The step index to use for next data
+  /// \param data The vector of pointer and size of the expected outputs
+  /// \return cb::Error object indicating success or failure.
+  cb::Error UpdateValidationOutputs(
+      const std::vector<const cb::InferRequestedOutput*>& outputs,
+      int stream_index, int step_index,
+      std::vector<std::vector<std::pair<const uint8_t*, size_t>>>& data);
+
+  /// Helper function to update the inputs
+  /// \param inputs The vector of pointers to InferInput objects for all
+  /// possible inputs, potentially including optional inputs with no provided
+  /// data
+  /// \param valid_inputs The vector of pointers to InferInput objects to be
+  /// used for inference request.
+  /// \param stream_index The data stream to use for next data
+  /// \param step_index The step index to use for next data
+  /// \return cb::Error object indicating success or failure.
+  cb::Error SetInputs(
+      const std::vector<cb::InferInput*>& inputs,
+      std::vector<cb::InferInput*>& valid_inputs, const int stream_index,
+      const int step_index);
+
+
+ protected:
```

Comment:
I personally tend to shy away from friend classes in general since the friend concept defeats access control. If used irresponsibly, it can allow for incorrect access outside of the containing class. Protected enables it to be used by child classes but nothing else, which is still a tighter access control mechanism. If we want to go another way, i am happy to but i think we are talking about coding style/practices that as a team needs to be discussed in depth more.  
I think this worth a discussion though. Thanks for pointing it out.